File: webappd/src/main/java/webappd/DataHolder.java
package webappd;

import lombok.Synchronized;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.OutputStream;
import java.util.zip.GZIPInputStream;
import java.util.zip.GZIPOutputStream;

class DataHolder {
    final static String PORT_URI = "/kcsapi/api_port/port";
    final static String START_URI = "/kcsapi/api_start2/getData";

    private static DataHolder INSTANCE = new DataHolder();

    private byte[] compressedStartJSON = null;
    private String portJSON = null;

    private DataHolder() {
    }

    @Synchronized
    String getPortJSON() {
        return this.portJSON;
    }

    @Synchronized
    void setPortJSON(String jsonStr) {
        this.portJSON = jsonStr;
    }

    @Synchronized
    String getStartJSON() {
        if (this.compressedStartJSON == null) {
            return null;
        }
        try {
            return this.gzDecompress(this.compressedStartJSON);
        } catch (IOException e) {
            e.printStackTrace();
            return null;
        }
    }

    @Synchronized
    void setStartJSON(String jsonStr) {
        try {
            this.compressedStartJSON = this.gzCompress(jsonStr);
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    private byte[] gzCompress(String str) throws IOException {
        ByteArrayOutputStream out = new ByteArrayOutputStream();
        try (OutputStream gzip = new GZIPOutputStream(out)) {
            gzip.write(str.getBytes());
            return out.toByteArray();
        }
    }

    private String gzDecompress(byte[] data) throws IOException {
        try (
                ByteArrayOutputStream out = new ByteArrayOutputStream();
                ByteArrayInputStream in = new ByteArrayInputStream(data);
                GZIPInputStream zin = new GZIPInputStream(in);
        ) {
            byte[] buf = new byte[8192];
            int read;
            while ((read = zin.read(buf)) > 0) {
                out.write(buf, 0, read);
            }
            return out.toString();
        }
    }

    static DataHolder getInstance() {
        return INSTANCE;
    }
}

File: webappd/src/main/java/webappd/DataHolder.kt
package webappd

import lombok.Synchronized
import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.io.IOException
import java.nio.charset.StandardCharsets
import java.util.zip.GZIPInputStream
import java.util.zip.GZIPOutputStream

internal class DataHolder private constructor() {

    private var compressedStartJSON: ByteArray? = null

    @get:Synchronized("portLock")
    @set:Synchronized("portLock")
    var portJSON: String? = null

    var startJSON: String?
        @Synchronized("startLock")
        get() {
            if (this.compressedStartJSON == null) {
                return null
            }
            try {
                return this.gzDecompress(this.compressedStartJSON!!)
            } catch (e: IOException) {
                e.printStackTrace()
            }
            return null
        }
        @Synchronized("startLock")
        set(jsonStr) = try {
            this.compressedStartJSON = this.gzCompress(jsonStr!!)
        } catch (e: IOException) {
            e.printStackTrace()
        }

    @Throws(IOException::class)
    private fun gzCompress(str: String): ByteArray {
        ByteArrayOutputStream().use { out ->
            GZIPOutputStream(out).use { gzip ->
                gzip.write(str.toByteArray(StandardCharsets.UTF_8))
                gzip.close() // GZIPOutputStream must be closed before getting the compression result!
                return out.toByteArray()
            }
        }
    }

    @Throws(IOException::class)
    private fun gzDecompress(data: ByteArray): String {
        ByteArrayOutputStream().use { out ->
            GZIPInputStream(ByteArrayInputStream(data)).use { gzip ->
                val buf = ByteArray(8192)
                var read = 0
                while ({read = gzip.read(buf); read}() != -1) {
                    out.write(buf, 0, read)
                }
                return String(out.toByteArray(), StandardCharsets.UTF_8)
            }
        }
    }

    companion object {
        const val PORT_URI = "/kcsapi/api_port/port"
        const val START_URI = "/kcsapi/api_start2/getData"

        val instance = DataHolder()
    }
}
